pkg/controllers/user/certsexpiration/certsexpiration.go

```diff
@@ -0,0 +1,169 @@
+		UserContxt:    userContext,
+	}
+
+	userContext.Management.Management.Clusters("").AddHandler(ctx, "certificate-expiration", c.sync)
+}
+
+func (c Controller) sync(key string, cluster *v3.Cluster) (runtime.Object, error) {
+	if key == "" || cluster == nil || cluster.DeletionTimestamp != nil {
+		return cluster, nil
+	}
+
+	if cluster.Spec.RancherKubernetesEngineConfig == nil {
+		return cluster, nil
+	}
+	certsExpInfo := map[string]v3.CertExpiration{}
+
+	cluster, err := c.ClusterLister.Get("", key)
+	if err != nil {
+		return cluster, err
+	}
+	certBundle, err := c.getClusterCertificateBundle(cluster.Name)
+	if err != nil {
+		return cluster, err
+	}
+	for certName, certObj := range certBundle {
+		info, err := getCertExpiration(certObj.CertificatePEM)
+		if err != nil {
+			logrus.Debugf("failed to get expiration date for certificate [%s] for cluster [%s]:%v", certName, key, err)
+			continue
+		}
+		certsExpInfo[certName] = info
+	}
+	if !reflect.DeepEqual(cluster.Status.CertificatesExpiration, certsExpInfo) {
+		cluster.Status.CertificatesExpiration = certsExpInfo
+		return c.ClusterClient.Update(cluster.DeepCopy())
+	}
+	return cluster, nil
+
+}
+
+func (c Controller) getClusterCertificateBundle(clusterName string) (map[string]pki.CertificatePKI, error) {
+	// cluster has a state file ?
+	currentState, err := getRKECurrentStateFromStore(c.ClusterStore, clusterName)
+	if err != nil {
+		return nil, err
+	}
+	if currentState != nil {
+		cleanCertificateBundle(currentState.CertificatesBundle)
+		return currentState.CertificatesBundle, nil
+	}
+
+	// No state file, let's try get the certs from the user cluster
+	// clusterContext, err := c.ClusterManager.UserContext(clusterName)
+	// if err != nil {
+	// 	return nil, err
+	// }
+	certs, err := getCertsFromUserCluster(c.UserContxt)
+	if err != nil {
+		return nil, err
+	}
+	cleanCertificateBundle(certs)
+	return certs, nil
+}
+
+func getRKECurrentStateFromStore(store cluster.PersistentStore, clusterName string) (*rkecluster.State, error) {
+	cluster, err := store.Get(clusterName)
+	if err != nil {
+		return nil, err
+	}
+	var fullState rkecluster.FullState
+	stateStr, ok := cluster.Metadata["fullState"]
+	if !ok {
+		return nil, nil
+	}
+	err = json.Unmarshal([]byte(stateStr), &fullState)
+	if err != nil {
+		return nil, err
+	}
+	return &fullState.CurrentState, nil
+}
+
+func getCertsFromUserCluster(c *config.UserContext) (map[string]pki.CertificatePKI, error) {
+	certs := map[string]pki.CertificatePKI{}
+	secrets, err := c.Core.Secrets("").Controller().Lister().List("kube-system", labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+	for _, s := range secrets {
+		if strings.HasPrefix(s.GetName(), "kube-") &&
+			s.Type == corev1.SecretTypeOpaque {
+			name := s.GetName()
+			secret, err := c.Core.Secrets("").Controller().Lister().Get("kube-system", name)
```

Comment:
question, don't we already have secret object since we listed them all on L118, or am I missing something?